mysql ex incremental

pivnicek opened the pull request:
Fixes #1533 

Proposed changes:

- use configuration rows for supported components
  - add create/update/delete configuration rows to the installedComponentStore
- allow selecting of `incrementalFetching` parameter for tables that qualify (qualifying tables are those with auto-increment id or timestamp columns)
- show last row fetched status for config rows that have incremental fetching
- have button available to migrate supported component configurations to config rows

pivnicek commented:
TODOs:

features:
- [ ] show last row fetched status for configs with incremental fetching enabled.
- [ ] allow edit of last row fetched for configs with incremental fetching enabled.
- [ ] display warnings appropriate to incrementalFetching column type selected.
- [ ] component migration button

bugs:
- [ ] fix saving new configrow on direct load of new query page

pivnicek commented:
With sections it seems maybe cleaner:
![image](https://user-images.githubusercontent.com/569139/38992331-b4486d64-43e0-11e8-8efc-bd98a6653b40.png)

ondrejhlavacek commented:
Got a few warnings, but cannot reproduce

![image](https://user-images.githubusercontent.com/497675/39047772-72d6eb24-449b-11e8-9025-16cfa5826fe0.png)

ondrejhlavacek commented:
This somehow magically changes

![image](https://user-images.githubusercontent.com/497675/39047851-b41101ce-449b-11e8-991b-db88a3d3fb12.png)

to

![image](https://user-images.githubusercontent.com/497675/39047868-c3736418-449b-11e8-938b-7540e3c8e383.png)

ondrejhlavacek commented:
Maybe along with it, or completely unrelated, got this error

![image](https://user-images.githubusercontent.com/497675/39047893-dd7fff10-449b-11e8-8c53-b4c0aebaba03.png)

ondrejhlavacek commented:
![image](https://user-images.githubusercontent.com/497675/39048292-3955ecd6-449d-11e8-91a7-2d625ca17e44.png)

Not too sure about this. It does not tell me about benefits. Maybe the migration component should take care of this?

ondrejhlavacek commented:
I can't get the `Incremental Fetching` show up anywhere. So what if the section is there all the time, but if it's not an incremental fetching candidate, all inputs will be disabled and a short note will explain what I have to do to be able to use it?

ondrejhlavacek commented on src/scripts/modules/ex-db-generic/react/components/QueryEditor.jsx:
```diff
@@ -315,6 +337,27 @@ export default React.createClass({
             />
           </div>
+          {this.renderIncrementalFetching()}
+          <h3>Loading Options</h3>
```
Nvm, all all other sections seem to have `ing`

pivnicek commented:
Thanks for taking a look @ondrejhlavacek .

wrt
> Not too sure about this. It does not tell me about benefits. Maybe the migration component should take care of this? I can't see any direct benefit or change in the config.

There isn't any visible difference to migrating the config other than it does then allow you to access the incremental fetching section (if the table contains an auto-increment id or timestamp)

There's no reason afaik to use the migration component (I thought about it).  The migration just takes each table from the config and puts it in a config-row.

pivnicek commented:
todo:

- [ ] fix error/weird behaviour on fresh config startup as reported above
- [ ] when db creds save it makes an empty table [] in the config, shouldn't do that

ujovlado commented on src/scripts/modules/ex-db-generic/react/components/MigrateToRowsButton.jsx:
```diff
@@ -0,0 +1,26 @@
+import React from 'react';
+
+export default React.createClass({
+  displayName: 'MigrateToRowsButton',
```
no need to have `displayName`

ujovlado commented on src/scripts/modules/ex-db-generic/react/components/MigrateToRowsButton.jsx:
```diff
@@ -0,0 +1,26 @@
+export default React.createClass({
+  displayName: 'MigrateToRowsButton',
+  propTypes: {
+    componentId: React.PropTypes.string,
```
Shouldnt be `componentId` required?

ujovlado commented on src/scripts/modules/ex-db-generic/storeProvisioning.js:
```diff
@@ -6,23 +6,77 @@ import hasSshTunnel from './templates/hasSshTunnel';
 
 const defaultSshPort = 22;
 
+export const rowConfigurationType = 'row';
```
Are this constants? I would prefer uppercase then.

ujovlado commented on src/scripts/modules/ex-db-generic/storeProvisioning.js:
```diff
@@ -6,23 +6,77 @@ import hasSshTunnel from './templates/hasSshTunnel';
+  let queryState = query.has('state') ? query.get('state').toJS() : {};
+  let paramsQuery = query.delete('state');
+  return {
+    'rowId': query.get('id'),
```
Why are you creating JSON here? Shouldn't this method return JS (or immutable) object and JSON.stringify called somewhere in "installed components action creators"?

pivnicek commented:
@ujovlado great!  thanks a lot!

Oh, I had no idea that it looked like that on linux/win, it's using the same markup as the s3 extractor (does that one look funny for you too?).  It's a pity, because imho it looks quite nice on the mac.

Regarding the lets, yes, it's just pure laziness on my part:)  I will go through and tighten it up where appropriate.  I'll also make the changes you suggest inline.

Cheers!

Btw, regarding the config migration button, what do you think, should it be more descriptive?

ujovlado commented:
@pivnicek regarding config migration (thanks for reminding), I was only missing some kind of success message. We need to provide feedback to user that migration was ok (or wasn't). `alert-success` that migration was successful? Anyway, I like how it's done and I can see each row in versions history

pivnicek commented:
ok @ujovlado, I updated it to reflect your feedback.
regarding the "look", I've made an indigo-ui issue for that here: https://github.com/keboola/indigo-ui/issues/155

ujovlado commented:
@pivnicek good, it's great to see result of migration ;)

One thing. I see you tried to show Loader if migration is pending. But I cannot see setting of `migration.pending` anywhere -> it didn't show Loader for me.

ujovlado commented on src/scripts/modules/ex-db-generic/react/components/MigrateToRowsButton.jsx:
```diff
@@ -0,0 +1,32 @@
+    ExDbActionCreators.migrateConfig(this.props.configId);
+  },
+
+  dismissAlert() {
```
@pivnicek this fn is not used

pivnicek commented:
@ujovlado yes, you're right, I've removed the loader, it's quick enough that it's not really necessary imho.  cheers.

ujovlado commented:
@ondrejhlavacek vidim, ze to mas este v svojom "Najlos TODO". Budes to chciet este checknut?

janmichek commented:
about that label / font size thing. This is the issue of platform rendering. This will probably looks different accross the platform, becasuse we are using Arial font (on Mac is rendered as  Helvetica).

On pc, looks pretty decent
![image](https://user-images.githubusercontent.com/15363559/39870700-92c3e866-5462-11e8-89a6-7dfecf5cfe4b.png)

I don't think we can tune it per platfrom. The ultimate solutio could be this https://github.com/keboola/indigo-ui/issues/27 I may escalate it

ujovlado commented:
I agree, it's platform specific. But if we choose to use 300/400 font-weight we have to be sure those weights are available on all platforms. Original idea was Roboto font (which never makes it to production due to typo). Let's continue in indigo: keboola/indigo-ui#155

ujovlado commented:
@pivnicek great, it's working as expected.

ondrejhlavacek commented:
@ujovlado dal bych tomu ještě pár prokliknutí

ondrejhlavacek commented:
when I click on save credentials, the form changes to a read-only mode for a moment. 

![image](https://user-images.githubusercontent.com/497675/39874268-9e0f1b9a-546d-11e8-92e5-7636b91d8f85.png)

http://take.ms/VIq5Q

is there any purpose of this? cannot see this form anywhere. some of the input fields move a bit, some change drastically (checkboxes). if the reason is to be read-only during save, `disabled` attribute is just enough. 

If there's a reason Network Compression does not have the nice checkbox and is missing help (in the edit form as well).

pivnicek commented:
@ondrejhlavacek  totally agree with you, I thought I had already made an issue for this, but I guess I hadn't.  https://github.com/keboola/kbc-ui/issues/1663 is now.

ondrejhlavacek commented:
![image](https://user-images.githubusercontent.com/497675/39875120-f402b4ec-546f-11e8-98d0-1354d763f1b0.png)

maybe without the ellipsis?

ondrejhlavacek commented:
testing credentials without editing/saving config does not remove warning. i cannot start adding tables even if the connection has been tested

http://take.ms/bBa7Q

ondrejhlavacek commented:
another ellipsis

![image](https://user-images.githubusercontent.com/497675/39875456-fa101c3e-5470-11e8-88fd-dce4e7efd927.png)

ondrejhlavacek commented:
connected to the db, selected 2 tables and clicked "create". this weirdo showed up

![image](https://user-images.githubusercontent.com/497675/39875608-5d28b0c4-5471-11e8-98e8-036b57cc819c.png)

ondrejhlavacek commented:
another ellipsis 

![image](https://user-images.githubusercontent.com/497675/39875698-a05bd02e-5471-11e8-9886-9eec529934b1.png)

ondrejhlavacek commented:
![image](https://user-images.githubusercontent.com/497675/39875721-b117fe74-5471-11e8-8a4f-a1d029fd4397.png)

`select tables to copy` => `select tables to import`? `select tables to extract`?

ondrejhlavacek commented:
i cant get to the table detail, sending connection details in pm

ondrejhlavacek commented:
this is the `getTables` API response

```json
{
    "tables": [
        {
            "name": "test",
            "schema": "keboola",
            "type": "BASE TABLE",
            "rowCount": "0",
            "autoIncrement": "1",
            "columns": [
                {
                    "name": "id",
                    "sanitizedName": "id",
                    "type": "int",
                    "primaryKey": true,
                    "length": "10",
                    "nullable": false,
                    "default": null,
                    "ordinalPosition": "1",
                    "extra": "auto_increment",
                    "autoIncrement": "1"
                }
            ]
        },
        {
            "name": "test2",
            "schema": "keboola",
            "type": "BASE TABLE",
            "rowCount": "0",
            "autoIncrement": "1",
            "columns": [
                {
                    "name": "id",
                    "sanitizedName": "id",
                    "type": "int",
                    "primaryKey": true,
                    "length": "10",
                    "nullable": false,
                    "default": null,
                    "ordinalPosition": "1",
                    "extra": "auto_increment",
                    "autoIncrement": "1"
                },
                {
                    "name": "col1",
                    "sanitizedName": "col1",
                    "type": "int",
                    "primaryKey": false,
                    "length": "10",
                    "nullable": true,
                    "default": null,
                    "ordinalPosition": "2"
                }
            ]
        },
        {
            "name": "test3",
            "schema": "keboola",
            "type": "BASE TABLE",
            "rowCount": "0",
            "columns": [
                {
                    "name": "id",
                    "sanitizedName": "id",
                    "type": "varchar",
                    "primaryKey": false,
                    "length": "11",
                    "nullable": false,
                    "default": "",
                    "ordinalPosition": "1"
                }
            ]
        }
    ],
    "status": "success"
}
```

ondrejhlavacek commented:
clicking on `create` button throws an error

![image](https://user-images.githubusercontent.com/497675/39876382-4d251152-5473-11e8-9693-7721b5f74d0c.png)

pivnicek commented:
ok @ondrejhlavacek thanks for spotting that bug!  it is resolved now.

ondrejhlavacek commented:
Configured, ran and got this

![image](https://user-images.githubusercontent.com/497675/39878300-4f1f2e02-5478-11e8-9664-786186736418.png)

Did not check the network compression checkbox though.

ondrejhlavacek commented:
![image](https://user-images.githubusercontent.com/497675/39878519-e3257502-5478-11e8-919f-41844c61a57f.png)

remove `Note: `

pivnicek commented:
Nice one, backend bug, there is no NetworkCompression property in https://github.com/keboola/db-extractor-mysql/blob/master/src/Keboola/DbExtractor/Configuration/MySQLConfigRowDefinition.php

ondrejhlavacek commented:
Incremental loading mysteriously shows up after a while. I know it's because there's sync actions in the background, but it surprised me.

http://take.ms/QbqQV

I think this section should be visible at all times even for tables that cannot be loaded incrementally with a note what are the requirements for incremental loading. In the recorded case even a disabled state is probably not required, or ALL input fields should be disabled while loading table details.

In case the table which has incremental loading set up is deleted from the db (or is not visible using the credentials), the inremental loading section never shows up.

pivnicek commented:
@ondrejhlavacek Good point there, I will update it as you say.  thanks!

ondrejhlavacek commented:
Otherwise looks good! I have one question/comment. The relation between Incremental Loading and Incremental Fetching is somewhat hidden and maybe counterintuitive. I can setup incremental fetching without incremental loading. I knew what I was doing (testing it), but I guess this is not the regular use case. Not sure what to do about it though. I'd probably fancy a warning sayint that incremental fetching is turned on but incremental loading is not and explain the consequences.

![image](https://user-images.githubusercontent.com/497675/39880311-b7504966-547d-11e8-8022-42b4d846264a.png)

ondrejhlavacek commented:
![image](https://user-images.githubusercontent.com/497675/39880483-3b820a44-547e-11e8-9ced-31aedf6277cc.png)
Remove `Note: `

pivnicek commented:
@ondrejhlavacek thanks so much for the feedback.

- [ ]  remove `Note`
- [ ]  don't completely hide incremental fetching section while loading tables, just leave it disabled
- [ ]  if the component supports incremental fetching, but the selected table doesn't qualify, leave the section there with some text explaining why it is disabled.
- [ ]  if incremental fetching is ON and incremental loading is OFF put there some help text explaining what the behaviour will be
- [ ] edit state text ("state" is probably meaningless to the user)
- [ ] Move state button to newline and add help text

I'll mark the comments above with a 👍 for fixed and a ❤️ for those that I've moved to #1663 
Cheers!

pivnicek commented:
ok @ondrejhlavacek I've updated it to reflect your points (Hope I haven't forgotten anything!)

ondrejhlavacek commented:
![image](https://user-images.githubusercontent.com/497675/40060420-d55128f4-5856-11e8-8561-254889933cb3.png)

empty modal

ondrejhlavacek commented:
getting this warning in console

![image](https://user-images.githubusercontent.com/497675/40060488-0d3d80a0-5857-11e8-833d-6bf6ffedfec3.png)

ondrejhlavacek commented:
**Reset state**

I think there's no need to stage state reset. Currently resetting state changes the Saved button to Save, if it's a confirming modal and it's not a configuration, I would save it directly. It must be a pain to save both local state and configuration and handle it! Config rows components reset state directly after confirming.

ondrejhlavacek commented:
Getting a weird horizontal scrollbar (wired mouse connected, which shows all scrollbars all the time)

![image](https://user-images.githubusercontent.com/497675/40060753-9ad750da-5857-11e8-9477-e7c97ac87052.png)

but it's also present in the current ui

![image](https://user-images.githubusercontent.com/497675/40061017-40866494-5858-11e8-90fd-0b64f9c90a24.png)

so probably => https://github.com/keboola/kbc-ui/issues/1663

ondrejhlavacek commented:
Looks like there's an extra space character before `Clear`, the space between the icon an the text looks quite large.

![image](https://user-images.githubusercontent.com/497675/40061172-9de5187e-5858-11e8-9a66-b8d3e9d28248.png)

Also I'd probably run this UI element through #uiux, looks similar to the autoload datatypes?

ondrejhlavacek commented:
Also got randomly hit by 

![image](https://user-images.githubusercontent.com/497675/40061286-dce794ac-5858-11e8-8d40-1bfe4d73971d.png)

Not too sure how to reproduce it though :-(

pivnicek commented:
> Also I'd probably run this UI element through #uiux, looks similar to the autoload datatypes?

It's the same `reset state` component you're using in config rows, I just provided a label

ondrejhlavacek commented:
> > Also I'd probably run this UI element through #uiux, looks similar to the autoload datatypes?

> It's the same `reset state` component you're using in config rows, I just provided a label

but it's used in a different context, so it may be a component with a different flow/ui? 

![image](https://user-images.githubusercontent.com/497675/40061657-bcf5863a-5859-11e8-86f9-3ccfa09262c6.png)

pivnicek commented:
> but it's used in a different context, so it may be a component with a different flow/ui? the used component is supposed to be used in the right bar

yeah, that probably explains the weird looking spacing issue.  ok, I'll rewrite it to use something specific for this case.  👍

pivnicek commented:
Todo:

- refactor "reset state" button
- rebase :-)

pivnicek commented:
@ondrejhlavacek I believe I addressed the last batch of comments.  I turned the reset state thing into a simple link.  I've kept it as being part of the save/reset flow though, so if you clear the value the `reset` link by save will recover it for you.  This makes sense to me for how the form is laid out.

@ujovlado I rebased the commits trying to make sure all the messages are sensible and making a couple `fixups`

ujovlado commented:
@pivnicek can we discuss this please?

![screenshot_2018-05-17_10-17-58](https://user-images.githubusercontent.com/419849/40165230-ba9e5c6e-59bb-11e8-9855-39d8b7c30fb5.png)

I'm a bit confused by that message. Should I enable incremental loading? I think that we should enable it automatically if I choose incremental fetching. Does it make sense to have incremental fetching w/o incremental load?

pivnicek commented:
@ujovlado 
> I'm a bit confused by that message. Should I enable incremental loading? I think that we should enable it automatically if I choose incremental fetching. Does it make sense to have incremental fetching w/o incremental load?

yes, there is a use case for having a table in storage containing only the last fetch group imho, but it is definitely an edge case.  But your suggestion for enabling incremental loading automatically when a fetching column is selected is a good one, thanks!  Though I think the user should be allowed to uncheck it (then this warning will appear).  Will update now, cheers!

ujovlado commented:
By confusion I meant that message about "incremental fetching enabled" is shown bellow "incremental loading" unchecked. I think if we change that sentence a bit it'll do the thing. Something like "It is recommended to have incremental loading enabled if you have selected column for incremental fetching" (but I'm not 100% sure ;) )

pivnicek commented:
Ah, sure.  how about:

"It is recommended to enable incremental loading if using incremental fetching.  If incremental loading is *not* enabled, the storage table will only ever contain the most recently fetched results"

ujovlado commented:
Yep, that's great :+1:

ondrejhlavacek commented:
- `Last fetched value` => `Last Fetched Value`

ondrejhlavacek commented on src/scripts/modules/ex-db-generic/react/components/QueryEditor.jsx:
```diff
@@ -477,6 +539,145 @@ export default React.createClass({
+    }
+  },
+
+  renderlastFetchedInfo() {
```
`renderLastFetchedInfo`

ondrejhlavacek commented:
One more thought to the reset state. How about this?

![image](https://user-images.githubusercontent.com/497675/40168354-8ec82be4-59c3-11e8-81d7-b1f5fa15af0b.png)

```
          <div className="form-control-static">
            <div>
              <strong>{lastFetchedRowValue}</strong>
              {' '}
              <a onClick={this.handleStateReset}>
                <Tooltip
                  tooltip={tooltip}
                  placement="top"
                >
                  <span>
                    <i className="fa fa-fw fa-undo"/>
                    Reset
                  </span>
                </Tooltip>
              </a>
            </div>
            <div className="help-block">
              The highest value of {fetchingColumn} in the latest successful run.
            </div>
          </div>
```

pivnicek commented:
@ondrejhlavacek I dunno.  To me `clear` is better than `reset`.  `reset` to me means return to previous value, whereas `clear` means goto 0.

ujovlado commented:
@pivnicek I want to merge this but found one "glitch".

How to reproduce:

1. wait for "Fetching table list from source database ..." in old configuration format (e.g. in query detail)
2. go back to index page
3. migrate configuration
4. go back to query detail
5. incremental fetching is not available for me

if I "Reload" tables again it will appear.

What do you think? Is it solvable?

pivnicek commented:
@ujovlado Ah, yes, because it only looks for candidate tables if the configuration is in the new format and it looks for these tables when it fetches them.

I suppose we could trigger a re-fetch of the tables upon config migration

pivnicek commented:
I'll do it now, will take just a few minutes.  Thanks for catching that

ujovlado commented:
> I suppose we could trigger a re-fetch of the tables upon config migration

Is it needed? I suppose it's in state somewhere.

So having pseudo fn "is incremental fetching available" will return false for old state, but after migration state should change and function will return true and it'll available :)

pivnicek commented:
hmmm, it shouldn't be necessary, you're right, it was just the quickest dirtiest way I could think of:)

Let me have a look, I'll report back shortly.

pivnicek commented:
ok @ujovlado I've fixed it without requiring any extra requests.  Pls let me know what you think.

ujovlado commented:
@pivnicek still not ready ... I fixed two additional issues there (see last 2 commits).  I think we should prevent call `getCandidateTable` if query has no `table` (currently I hotfixed it by returning `null`.

ujovlado commented:
Another issue is that switching to advanced mode doesn't reset incremental fetching options and extractor fails with err `Incremental fetching is not supported for advanced queries.`

pivnicek commented:
Thanks @ujovlado for being so thorough, I will fix the remaining issue and see if there is maybe a better option for the returning `null` hotfix for getCandidateTable.

pivnicek commented:
@ujovlado I've addressed the issues in the above commits.  Also I found that the fetching column should be reset on source table change so I added that too.

ujovlado commented:
@pivnicek migration stopped working :)

ujovlado commented:
![screenshot_2018-05-18_14-59-32](https://user-images.githubusercontent.com/419849/40235932-1b5b8928-5aac-11e8-9b00-7fb95c64f833.png)

pivnicek commented:
ahhh, damn yeah, if you migrate before tables are fetched it will do that omg.  mmt

pivnicek commented:
ok, @ujovlado, omg that should do it
